src/decima/utils/inject.py

```diff
@@ -0,0 +1,226 @@
+from grelu.sequence.utils import reverse_complement
+from grelu.sequence.format import intervals_to_strings
+
+from decima.core.metadata import GeneMetadata
+
+
+class SeqBuilder:
+    """
+    Build the sequence from the variants.
+
+    Args:
+        chrom: chromosome
+        start: start position
+        end: end position
+        anchor: anchor position
+        track: track positions shifts due to indels.
+    """
+
+    def __init__(self, chrom: str, start: int, end: int, anchor: int, track: List[int] = None):
+        self.chrom = chrom
+        self.start = start
+        self.end = end
+        self.variants = list()
+        self.anchor = anchor
+        self.start_shift = 0  # how much interval is shifted to the left upstream
+        self.end_shift = 0  # how much interval is shifted to the right downstream
+        self.shifts = {pos: 0 for pos in track or list()}
+
+    @staticmethod
+    def _split_variant(variant, pos):
+        start = variant["pos"]
+        end = start + len(variant["ref"])
+
+        assert start <= pos < end, f"Variant position {variant['pos']} is out of range [{start}, {end})"
+        _pos = pos - start
+
+        left_variant = {
+            "chrom": variant["chrom"],
+            "pos": start,
+            "ref": variant["ref"][:_pos],
+            "alt": variant["alt"][:_pos],
+        }
+        right_variant = {
+            "chrom": variant["chrom"],
+            "pos": pos,
+            "ref": variant["ref"][_pos:],
+            "alt": variant["alt"][_pos:],
+        }
+        return left_variant, right_variant
+
+    def inject(self, variant: Dict):
+        """
+        Inject the variant into the sequence.
+
+        Args:
+            variant: variant to inject in the format of {"chrom": str, "pos": int, "ref": str, "alt": str}
+
+        Returns:
+            self
+        """
+        variant = dict(variant)
+        variant["ref"] = variant["ref"].replace(".", "")
+        variant["alt"] = variant["alt"].replace(".", "")
+
+        if variant["chrom"] != self.chrom:
+            warnings.warn(f"Variant chromosome `{variant['chrom']}` does not match `{self.chrom}`. Skipping...")
+            return self
+
+        for i in self.variants:
+            if variant["pos"] == i["pos"]:
+                raise ValueError(
+                    f"At this position `{variant['pos']}` there is already a variant `{i}` "
+                    f"thus cannot inject `{variant}` variant at the same position."
+                    " Please check variant and ensure not redundant positions."
+                )
+
+        variant_start = variant["pos"]
+        variant_end = variant_start + len(variant["ref"])
+
+        if variant_end <= self.start:
+            warnings.warn(
+                f"Variant position `{variant['pos']}` is upstream of the interval `[{self.start}, {self.end}]`. Skipping..."
+            )
+            return self
+        elif variant_start < self.start < variant_end:
```

Comment:
If I understand right, this is a case in which only part of the variant overlaps with the interval bounds - why should we allow this at all? If we are only able to include half of our variant in the input sequence, we won't be able to make an accurate prediction for it, right?

Reply:
splitting variants removed for boundaries